diff --git a/src/dawn/native/vulkan/SwapChainVk.cpp b/src/dawn/native/vulkan/SwapChainVk.cpp
index 41447ee57..e74899698 100644
--- a/src/dawn/native/vulkan/SwapChainVk.cpp
+++ b/src/dawn/native/vulkan/SwapChainVk.cpp
@@ -69,6 +69,22 @@ VkPresentModeKHR ToVulkanPresentMode(wgpu::PresentMode mode) {
     DAWN_UNREACHABLE();
 }
 
+VkCompositeAlphaFlagBitsKHR ToVulkanCompositeAlphaMode(wgpu::CompositeAlphaMode mode) {
+    switch (mode) {
+        case wgpu::CompositeAlphaMode::Opaque:
+            return VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR;
+        case wgpu::CompositeAlphaMode::Premultiplied:
+            return VK_COMPOSITE_ALPHA_PRE_MULTIPLIED_BIT_KHR;
+        case wgpu::CompositeAlphaMode::Unpremultiplied:
+            return VK_COMPOSITE_ALPHA_POST_MULTIPLIED_BIT_KHR;
+        case wgpu::CompositeAlphaMode::Inherit:
+            return VK_COMPOSITE_ALPHA_INHERIT_BIT_KHR;
+        case wgpu::CompositeAlphaMode::Auto:
+            break;
+    }
+    DAWN_UNREACHABLE();
+}
+
 uint32_t MinImageCountForPresentMode(VkPresentModeKHR mode) {
     switch (mode) {
         case VK_PRESENT_MODE_FIFO_KHR:
@@ -296,33 +312,16 @@ ResultOrError<SwapChain::Config> SwapChain::ChooseConfig(
             "Vulkan SwapChain must support %s with sRGB colorspace.", config.wgpuFormat));
     }
 
-    // Only the identity transform with opaque alpha is supported for now.
+    // Only the identity transform is supported for now.
     DAWN_INVALID_IF(
         (surfaceInfo.capabilities.supportedTransforms & VK_SURFACE_TRANSFORM_IDENTITY_BIT_KHR) == 0,
         "Vulkan SwapChain must support the identity transform.");
 
     config.transform = VK_SURFACE_TRANSFORM_IDENTITY_BIT_KHR;
 
-    config.alphaMode = VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR;
-#if !DAWN_PLATFORM_IS(ANDROID)
+    config.alphaMode = ToVulkanCompositeAlphaMode(GetAlphaMode());
-    DAWN_INVALID_IF(
-        (surfaceInfo.capabilities.supportedCompositeAlpha & VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR) == 0,
-        "Vulkan SwapChain must support opaque alpha.");
-#else
-    // TODO(dawn:286): investigate composite alpha for WebGPU native
-    std::array<VkCompositeAlphaFlagBitsKHR, 4u> compositeAlphaFlags = {
-        VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR,
-        VK_COMPOSITE_ALPHA_PRE_MULTIPLIED_BIT_KHR,
-        VK_COMPOSITE_ALPHA_POST_MULTIPLIED_BIT_KHR,
-        VK_COMPOSITE_ALPHA_INHERIT_BIT_KHR,
-    };
-    for (uint32_t i : Range(4u)) {
-        if (surfaceInfo.capabilities.supportedCompositeAlpha & compositeAlphaFlags[i]) {
-            config.alphaMode = compositeAlphaFlags[i];
-            break;
-        }
-    }
-#endif  // #if !DAWN_PLATFORM_IS(ANDROID)
+    DAWN_INVALID_IF((surfaceInfo.capabilities.supportedCompositeAlpha & config.alphaMode) == 0,
+                    "Vulkan SwapChain does not support the requested composite alpha mode.");
 
     // Choose the number of images for the swapchain= and clamp it to the min and max from the
     // surface capabilities. maxImageCount = 0 means there is no limit.
